Extended performance tips about abstract types

HenriDeh opened the pull request:
Following issue [#34739](https://github.com/JuliaLang/julia/issues/34739), I added examples where fields that are not abstract are also detrimental to performance. Specifically, fields with unparameterized types should also be avoided.

HenriDeh committed:
Extended performance tips about abstract types

Following issue [#34739](https://github.com/JuliaLang/julia/issues/34739), I added examples where fields that are not abstract are also detrimental to performance. Specifically, fields with unparameterized types should also be avoided.

fix doctest_linux64

Apparently buildbot/doctest_linux64 makes a difference between `Arrray{Float64,1}` and `Vector{Float64}`. Other buildbots don't and using `Array{Int64,1}` above does not seem to be a problem...

Reverse previous commit

Previous commit broke even more CI tests.

Change docstring output

Update performance-tips.md

Update performance-tips.md

StefanKarpinski commented:
I think that adding a whole separate section for this is unnecessary, it could just be noted that parametric types without all parameters bound are not concrete, so performance issues apply to them as well and that therefore one wants to either specify concrete values for parameters or path the parameters through to the containing type.

HenriDeh commented:
In this case, I changed "abstract" to "non-concrete" in the title of the section. I kept the example, if you feel like it's really unnecessary and that a single sentence/paragraph should be enough then I'll try to find one.

johnnychen94 commented on doc/src/manual/performance-tips.md:
````diff
@@ -331,7 +331,37 @@ For reasons of length the results are not shown here, but you may wish to try th
+           mpt::MyParametricType
+       end
+```
+Even though `MyParametricType` is not an abstract type, when it is not parameterized, it is not a concrete type either:
````
> `when it is not parameterized`

I read four times until I realized this belongs to `it is not a concrete type either` 😂  Perhaps you can adjust it a bit for nonnative English speakers.

HenriDeh replied:
I guess this is because I'm not native either. I changed this to something clearer.

timholy commented:
I think the abstract type is simpler than a not-fully-paramterized type. Consequently you should flip the order, keep the old material on abstract types and then add that not-full-parametrized types behave similarly. Like Stefan, I kind of think a single short example would suffice. Instead of inventing new types like `MyParametricType`, how about just distinguish `Array` and `Array{T}` from `Array{T,N}`?

johnnychen94 reviewed (commented):
Out of curiosity. If `MyParametricType` is a container like `AbstractArray`, will there be any performance difference for `MySectionType` and `MyThirdType`? IIUC `MySectionType` would generate too many concrete types compared to `MyThirdType` and thus be slower.

HenriDeh commented:
Okay thank you. I think this last commit is much better then.

StefanKarpinski commented on doc/src/manual/performance-tips.md:
```diff
@@ -331,6 +331,14 @@ For reasons of length the results are not shown here, but you may wish to try th
 the type is fully-specified in the first case, the compiler doesn't need to generate any code
 to resolve the type at run-time. This results in shorter and faster code.
 
+One should also keep in mind that not-fully-parameterized types behave like abstract types. For example, even though `Array` is not an abstract type, it is not a concrete type either when its parameters are not bound:
```
```suggestion
One should also keep in mind that not-fully-parameterized types behave like abstract types. For example, even though a fully specified `Array{T,n}` is concrete, `Array` itself with no parameters given is not concrete:
```

StefanKarpinski commented:
The terminology is a bit fuzzy here. There's nominal abstract types, i.e. the ones you explicitly declare with `abstract type A end`. However, I would call `Array` an abstract type as well. I edited a bit to avoid saying that `Array` is not an abstract type. True, it wasn't declared with `abstract type`, but it's not concrete, so it must be abstract. Man, the word "abstract" got real weird.

crstnbr commented:
> However, I would call `Array` an abstract type as well.

Well, I guess you could call it *abstract* but it isn't an abstract type (`isabstract(Array) == false`), as you noted. It's a `UnionAll` since it has free parameters. I like the idea of distinguishing the two cases: an abstract type is a **specific** *abstract* type (a node in the type tree) whereas anything with a free parameter is an *abstract* "**set of types**". Both are *abstract*, they are not concrete, but only one is an abstract type.

But I do agree that this is getting weird, in particular for beginners reading the manual, and I think the current way of circumventing this point by talking about non-concreteness is a good idea..

vtjnash commented on doc/src/manual/performance-tips.md:
````diff
@@ -331,6 +331,14 @@ For reasons of length the results are not shown here, but you may wish to try th
+One should also keep in mind that not-fully-parameterized types behave like abstract types. For example, even though a fully specified `Array{T,n}` is concrete, `Array` itself with no parameters given is not concrete:
+
+```jldoctest myambig3
+julia> isconcretetype(Array), isabstracttype(Array), isstructtype(Array), isconcretetype(Array{Int}), isconcretetype(Array{Int,1})
+(false, false, true, false, true)
````
```suggestion
julia> !isconcretetype(Array), !isabstracttype(Array), isstructtype(Array), !isconcretetype(Array{Int}), isconcretetype(Array{Int,1})
(true, true, true, true, true)
```

HenriDeh replied:
I'm not sure about this, it seems a bit harder to decode for the reader imho.

vtjnash replied:
My think was that it would be helpful that all of these are testing true, so the reader doesn't need to take the negation in their head

rfourquet replied:
In this case, to emphasize the all-true-tests, perhaps:
```julia
julia> !isconcretetype(Array) && !isabstracttype(Array) && isstructtype(Array) && !isconcretetype(Array{Int}) && isconcretetype(Array{Int,1})
true
```
?

HenriDeh replied:
It's just that I find `false` more readable than a `!` in front of a function.

HenriDeh commented:
By the way, I did not forget this PR, I just don't know what's missing for it to be merged.

StefanKarpinski commented:
Added triage to force eyeballs on this and a decision. Might just be a matter of merging it.

HenriDeh commented:
Since we're here. Any idea why some tests are failing? I'm not very familiar with CI.

StefanKarpinski commented:
I'm not sure what's going on with CI. I would try rebasing this PR onto current master and force pushing the PR.

HenriDeh commented:
Sorry I don't have much experience with git and collaborative development. I have my fork opened in gitkraken, what is it exactly that I should do ? I can also just make a new PR

StefanKarpinski commented:
I went ahead and did it for you. Hopefully CI passes and we can just merge 😁

HenriDeh commented:
Great thanks, there's still the triage though, unless we decide to leave it as it is because it is not that important.

StefanKarpinski commented:
At least with a rebase and green CI it will be mergable on the spot if it's approved.

HenriDeh commented:
buildbot/doctest_linux64 was failing because it was expecting `Vector{Float64}` instead of an `Array{Float64,1}`. That's weird, especially because the `Int64` example right above did not fail. All other tests were successful (see the details of commit https://github.com/JuliaLang/julia/pull/34946/commits/a9165f11829b0785f477d9cc786bda1dc2aba0d6).

Thus I changed Array to Vector and it was buildkite that failing (because of a missing key in .git...). see https://github.com/JuliaLang/julia/pull/34946/commits/c2f9c991dc78442ac8ad0f65e3672f1d954046b6

I reverted the change from Vector to Array in https://github.com/JuliaLang/julia/pull/34946/commits/a0c17c696f680066ca32c77971ac7aa970432253 to be in the exact same state as https://github.com/JuliaLang/julia/pull/34946/commits/a9165f11829b0785f477d9cc786bda1dc2aba0d6 but buildkite still fails. 

So I really don't know what's up with these tests.

JeffBezanson commented:
Triage approves; please rebase and adjust the doctest output to match the current printing.

HenriDeh commented:
I squashed all commits together. Checks passed before that so they should pass now as well. Then it's ready to merge \o/
